lib/archethic/beacon_chain/network_coordinates.ex

```diff
@@ -0,0 +1,193 @@
+defmodule Archethic.BeaconChain.NetworkCoordinates do
+  @moduledoc """
+  Network coordinates is a way to map latency between nodes and used to determine the closest nodes
+  """
+
+  @digits ["F", "E", "D", "C", "B", "A", "9", "8", "7", "6", "5", "4", "3", "2", "1", "0"]
+
+  @type latency_patch :: {String.t(), String.t()}
+
+  @doc """
+  Compute the network patch based on the matrix latencies
+
+  The matrix must be an Hollow matrix, symmetric and all the latencies must be positive
+
+  In physics, the center of mass for a distribution of masses in the space is the unique point where
+  the weighted relative position of the distributed mass sums to zero. Here, the sum of vectors to
+  all nodes from the center of mass is zero (All the nodes being the unit of mass)
+
+  The distance between center of mass and each node is then reduced to this formula using law of cosine
+  and center of mass:
+  `$$ dic^2 =  {1\over 2} \sum_{j=1}^{n} dij^2 - {1\over 2}\sum_{j=2}^{n}\sum_{k=1}^{j-1} djk^2 $$`
+
+  Then using the laws of Cosine, we can compute a gram matrix which would help to plot the nodes in a 2D plan:
+  `$$ gij =  {1\over 2} (dic^2 + djc^2 - dij^2) $$`
+
+  Finally, because the nodes are in a 2D plan, we can factorize the gram matrix to get the eigenvalues and
+  eigenvectors to find their corresponding coordinates
+
+  ### Examples
+
+      iex> NetworkCoordinates.get_patch_from_latencies(Nx.tensor([
+      ...>  [0, 100, 150],
+      ...>  [100, 0, 200],
+      ...>  [150, 200, 0]
+      ...> ], names: [:line, :column], type: {:f, 64}))
+      [
+        {"B", "8"},
+        {"0", "8"},
+        {"C", "8"}
+      ]
+  """
+  @spec get_patch_from_latencies(Nx.Tensor.t()) :: list(latency_patch())
+  def get_patch_from_latencies(matrix) do
+    center_mass = compute_distance_from_center_mass(matrix)
+    gram_matrix = get_gram_matrix(matrix, center_mass)
+    {x, y} = get_coordinates(gram_matrix)
+    get_patch_digits(x, y)
+  end
+
+  # defp get_matrix_tensor() do
+  #  "distance_matrix.dat"
+  #  |> File.read!
+  #  |> String.split("\n", trim: true)
+  #  |> Enum.map(fn line ->
+  #    line
+  #    |> String.split(",", trim: true)
+  #    |> Enum.map(&String.to_float/1)
+  #  end)
+  #  |> Nx.tensor(names: [:line, :column], type: {:f, 64})
+  # end
+
+  defp compute_distance_from_center_mass(tensor) do
+    matrix_size = Nx.size(tensor[0])
+
+    a =
+      tensor
+      |> Nx.power(2)
+      |> Nx.sum(axes: [:column])
+      |> Nx.multiply(Nx.tensor(1.0 / matrix_size, type: {:f, 64}))
+
+    excluded_first_row_tensor = tensor[1..-1//1]
+
+    b =
+      1..(matrix_size - 1)
+      |> Enum.map(fn i ->
+        excluded_first_row_tensor
+        |> Nx.slice([i - 1, 0], [1, i])
+        |> Nx.power(2)
+        |> Nx.sum()
+      end)
+      |> Nx.stack()
+      |> Nx.sum()
+      |> Nx.to_number()
```

Comment:
it may be worth it to precise that that the indexes start at 0 vs 1 in the formula

Reply:
Usually, algorithm start indexes a 1, while programming language start indexes at 0. I guess this is a standard way of thinking.